pkg/util/tracing/tracer.go

```diff
@@ -413,6 +416,20 @@ func (t *Tracer) startSpanGeneric(
 			// Local root span - put it into the registry of active local root
 			// spans. `Span.Finish` takes care of deleting it again.
 			t.activeSpans.Lock()
+
+			// Ensure that the registry does not grow unboundedly in case there
+			// is a leak. When the registry reaches max size, each new span added
+			// kicks out some old span. We rely on map iteration order here to
+			// make this cheap.
+			if toDelete := len(t.activeSpans.m) - maxSpanRegistrySize + 1; toDelete > 0 {
```

Comment:
Do you reckon dropping just the one (oldest) span when we've capped out could cause contention
when we've gone overboard? I assume we'll want to do more profiling again, but that seems possible to me.

Reply:
What do you mean? Deleting from the map (especially since it's bounded-size) should be approximately free.

Reply:
Seems cheap:

```
func BenchmarkMapDeleteAdd1(b *testing.B) {
	const size = 5000
	m := make(map[int]int, size)
	for i := 0; i < size; i++ {
		m[rand.Intn(10000*size)] = 123
	}
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		for idx := range m {
			//delete(m, idx)
			break
			_ = m[idx]
		}
		//m[i] = 456
	}
}
```

With the delete and assignment in, ~200ns/op. As quoted above, ~45ns/op. So the deletion+insertion dominate over doing the one-element iteration.

Reply:
Drive-by comment since I had to rebase on top of this code: the other option is to use a `cache.UnorderedCache` which is an LRU cache. Iterations don't count as accesses and neither do gets if you use `StealthyGet`, so only `Add`s will, which means the cache will evict the spans that are most likely to be leaks.